### PR TITLE
check validity of invite before accepting

### PR DESCRIPTION
**Pull request by jh-bate**
- check the validity of invite before accepting
- refactor tests a little so I could see what was going on

### CODE REVIEW

**Comment by jh-bate** [comment]
[[Potential privacy breach] Attacker can invite themselves to someone's Care Team if they know their User ID](https://trello.com/c/5TEv87bO/32-potential-privacy-breach-attacker-can-invite-themselves-to-someones-care-team-if-they-know-their-user-id)

**Comment by jh-bate** [comment]
[[Privilege Escalation] Attacker can add upload permissions to a read-only Care Team invitation](https://trello.com/c/XULIbu2g/31-privilege-escalation-attacker-can-add-upload-permissions-to-a-read-only-care-team-invitation)

File: api/invite.go [hunk 1]
```diff
@@ -144,70 +145,106 @@ func (a *Api) GetSentInvitations(res http.ResponseWriter, req *http.Request, var
...
-		if conf, err := a.findExistingConfirmation(accept, res); err != nil {
-			log.Printf("AcceptInvite: finding [%s]\n", err.Error())
+		conf, err := a.findExistingConfirmation(accept, res)
+		if err != nil {
+			log.Printf("AcceptInvite error while finding confirmation [%s]\n", err.Error())
 			a.sendModelAsResWithStatus(res, err, http.StatusInternalServerError)
 			return
-		} else if conf != nil {
+		}
+		if conf != nil {
```
Comment (darinkrauss) [inline api/invite.go]: What happens of `conf == nil`? Looks like it just returns without a response? Is that correct?

Also, I'd reverse this logic so the primary functionality is not in the `if` block. So:

```
if conf == nil {
  // Do something and return
}

// Everything else
```

File: api/invite.go [hunk 2]
```diff
@@ -144,70 +145,106 @@ func (a *Api) GetSentInvitations(res http.ResponseWriter, req *http.Request, var
...
+				a.sendModelAsResWithStatus(
+					res,
+					&status.StatusError{Status: status.NewStatus(http.StatusInternalServerError, STATUS_ERR_DECODING_CONFIRMATION)},
+					http.StatusInternalServerError,
+				)
+				return
+			}
+			log.Printf("AcceptInvite: permissions were set as [%v] after an invite was accepted", setPerms)
+			//we know the user now
+			//TODO - is this legit
```
Comment (darinkrauss) [inline api/invite.go]: TODO?

Comment (darinkrauss) [reply api/invite.go]: Why would you even need to do this? Did you just confirm this on line 204? Specifically `ValidateUserID(inviteeID, &validationErrors).`?

Comment (darinkrauss) [reply api/invite.go]: In fact, this is the "bug" where you are replacing the original invitation user with the incoming user specified in the request.

File: api/invite.go [hunk 3]
```diff
@@ -144,70 +145,106 @@ func (a *Api) GetSentInvitations(res http.ResponseWriter, req *http.Request, var
...
+			conf.UserId = inviteeID
+
+			conf.UpdateStatus(models.StatusCompleted)
+			if a.addOrUpdateConfirmation(conf, res) {
```
Comment (darinkrauss) [inline api/invite.go]: The typical Go workflow is to return early on failures, so I would reverse this logic:

```
if !a.addOrUpdateConfirmation(conf, res) {
		statusErr := &status.StatusError{Status: status.NewStatus(http.StatusNotFound, statusInviteNotFoundMessage)}
		log.Println("AcceptInvite ", statusErr.Error())
		a.sendModelAsResWithStatus(res, statusErr, http.StatusNotFound)
		return
}

a.logMetric("acceptinvite", req)
res.WriteHeader(http.StatusOK)
res.Write([]byte(STATUS_OK))
```

File: api/invite_test.go [hunk 4]
```diff
@@ -113,134 +112,145 @@ func TestDismissInvite_NoPerms(t *testing.T) {
 
 func TestInviteResponds(t *testing.T) {
 
-	tests := []toTest{
+	postResquestMethod := "POST"
```
Comment (darinkrauss) [inline api/invite_test.go]: If you want to replace the `"POST"` (et.al.) with constants then I think you should use the official ones. Specifically, in the `net/http` package, ones like `http.MethodPost`.

File: api/invite_test.go [hunk 5]
```diff
@@ -273,32 +283,53 @@ func TestInviteResponds(t *testing.T) {
...
+		if inviteTest.token != "" {
 			request.Header.Set(TP_SESSION_TOKEN, testing_token)
 		}
 		response := httptest.NewRecorder()
 		testRtr.ServeHTTP(response, request)
 
-		if response.Code != test.respCode {
-			t.Logf("TestId %d expected %d actual %d", idx, test.respCode, response.Code)
+		if response.Code != inviteTest.respCode {
+			t.Logf(
```
Comment (darinkrauss) [inline api/invite_test.go]: Not sure if this multi-line version really helps readability.

File: api/invite_test.go [hunk 6]
```diff
@@ -113,134 +112,145 @@ func TestDismissInvite_NoPerms(t *testing.T) {
...
 		{
-			// not found without the full path
-			method:   "PUT",
+			desc:     "request not found without the full path",
+			method:   putResquestMethod,
 			url:      "/accept/invite",
 			token:    testing_token_uid1,
 			respCode: http.StatusNotFound,
 		},
+		// TODO: don't have success path as would need to inject expetced confirmation
```
Comment (darinkrauss) [inline api/invite_test.go]: TODO?

File: api/invite.go (hunk 2 above)
Comment (jh-bate) [reply api/invite.go]: ahh - yeah I forgot that! I read the code in-hindsight and thought it made no sense

File: api/invite_test.go (hunk 4 above)
Comment (jh-bate) [reply api/invite_test.go]: agreed, it was late and I was searching for that

**Pull request merged**